tests/test_date.py

```diff
@@ -132,76 +204,171 @@ def test_periods(self, low, high, period, expected_results):
+        self.period = period
 
+    def when_date_intersecting_generated(self, dates):
+        self.intersected_dates = dates
 
-class ParseDateWithFormats(unittest.TestCase):
+    def then_date_arguments_and_date_range_length_should_equal(self):
+        self.assertEquals(1, len(self.intersected_dates))
+
+    def then_date_period_and_intersected_dates_should_equal(self):
+        self.assertEquals(self.period_low, self.intersected_dates[0])
+
+    def when_should_one_date_for_each_month(self):
+        self.intersected_dates = list(date.get_intersecting_periods(
+            self.period_low, self.period_high, period=self.period))
+
+    def when_empty_period(self):
+        self.intersected_dates = list(date.get_intersecting_periods(
+            self.period_low, self.period_high, period=self.period))
+
+    def then_should_one_date_for_each_month(self):
+        self.assertEquals(self.expected_results, self.intersected_dates)
+
+    def then_empty_period(self):
+        self.assertEquals([], self.intersected_dates)
+
+    def then_should_reject_easily_mistaken_dateutil_arguments(self):
+        with self.assertRaisesRegexp(ValueError, "Invalid period: {}".format(self.invalid_period)):
+            date.get_intersecting_periods(self.period_low, self.period, period=self.invalid_period).next()
+
+
+class ParseDateWithFormats(BaseTestCase):
+    def setUp(self):
+        super(ParseDateWithFormats, self).setUp()
+        self.dt_data = NotImplemented
 
     def test_shouldnt_parse_invalid_date(self):
-        self.assertIsNone(date.parse_with_formats('yesterday', ['%Y-%m-%d'])['date_obj'])
+        self.then_shouldnt_parse_invalid_date()
 
     def test_should_parse_date(self):
-        result = date.parse_with_formats('25-03-14', ['%d-%m-%y'])
-        self.assertEquals(datetime(2014, 3, 25).date(), result['date_obj'].date())
+        self.when_should_parse_date()
+        self.then_should_parse_date()
 
     def test_should_use_current_year_for_dates_without_year(self):
-        today = datetime.today()
-
-        result = date.parse_with_formats('09.16', ["%m.%d"])
-        self.assertEquals(datetime(today.year, 9, 16).date(), result['date_obj'].date())
+        self.given_current_year()
+        self.when_should_use_current_year_for_dates_without_year()
+        self.then_should_use_current_year_for_dates_without_year()
 
     def test_should_use_current_date_for_dates_without_day(self):
-        twelfth = datetime(2014, 8, 12)
-        datetime_mock = Mock(wraps=datetime)
-        datetime_mock.utcnow = Mock(return_value=twelfth)
+        self.given_date(datetime(2014, 8, 31).date())
+        self.given_twelfth_first()
+        self.given_datetime_mock_first()
+        self.when_patch_first()
+        self.then_current_year_for_dates_without_year_is_not_none()
+        self.then_period_for_current_year_is_equal()
+        self.then_check_if_dates_are_equal()
+        self.given_date(datetime(2014, 2, 28).date())
```

Comment:
This test describes single behavior repeated twice on different data. Could you refactor this please?